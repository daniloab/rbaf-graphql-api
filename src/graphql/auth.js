import jwt from 'jsonwebtoken'
import { User } from './model'
import { jwtSecret } from './config'

export async function getUser (token) {
  if (!token) return { user: null }

  try {
    const decodedToken = jwt.verify(token.substring(4), jwtSecret)

    const user = await User.findOne({ _id: decodedToken.id })

    return {
      user,
    }
  } catch (err) {
    return { user: null }
  }
}

export function generateToken (user) {
  return `JWT ${jwt.sign({ id: user._id }, jwtSecret)}`
}