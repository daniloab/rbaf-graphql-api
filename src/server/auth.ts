import jwt from "jsonwebtoken";
import { User } from "./models";
import { jwtSecret } from "./config";
import { IUser } from "./modules/user/UserModel";
import { ITeam } from "./modules/team/TeamModel";
import { getUser } from "./getUser";

export const auth = async (ctx, next) => {
  const { authorization, domainname } = ctx.header;

  const result = await getUser({ authorization, domainname });

  const { unauthorized, user, team } = result;

  if (unauthorized) {
    ctx.error = 401;
    ctx.body = {
      data: null,
      errors: [
        {
          message: "Invalid session",
          severity: "WARNING",
        },
      ],
    };
    return;
  }

  ctx.user = user;
  ctx.team = team;

  await next();
};

type GenerateTokenArgs = {
  user: IUser;
};

export const generateToken = ({ user }: GenerateTokenArgs) => {
  return `JWT ${jwt.sign({ id: user._id }, jwtSecret)}`;
};
