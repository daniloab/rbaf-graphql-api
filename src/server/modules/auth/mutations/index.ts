import UserRegisterEmailByEmail from "./UserRegisterEmailByEmailMutation";

export default {
  UserRegisterEmailByEmail,
};
