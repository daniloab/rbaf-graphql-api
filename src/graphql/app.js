require('dotenv').config()
import Koa from 'koa'
import GraphQLHTTP from 'koa-graphql'
import Router from 'koa-router'
import cors from '@koa/cors'
import bodyParser from 'koa-bodyparser'

import { getUser } from './auth';
import { schema } from './schema'

const app = new Koa()
const router = new Router();

const graphqlSettingsPerReq = async req => {
    const { user } = await getUser(req.header.authorization)

    return {
        graphiql: process.env.NODE_ENV !== 'production',
        schema,
        context: {
            user,
            req,
        },
        formatError: error => {
            console.log(error.message);
            console.log(error.locations);
            console.log(error.stack);

            return {
                message: error.message,
                locations: error.locations,
                stack: error.stack,
            };
        },
    }
}

const graphqlServer = GraphQLHTTP(graphqlSettingsPerReq)

router.all('/graphql', graphqlServer)

app.use(bodyParser())
app.use(cors())
app.use(router.routes()).use(router.allowedMethods());

export default app
// app.use(mount('/graphql', GraphQLHTTP({
//     schema: graphQLSchema,
//     rootValue: graphQLResolvers,
//     graphiql: true
// })));

// app.listen(GRAPHQL_PORT, () => console.log(
//     `GraphQL Server is now running on http://localhost:${GRAPHQL_PORT}/graphql`
// ))

// app.on('error', err => {
//     log.error('server error', err)
// })