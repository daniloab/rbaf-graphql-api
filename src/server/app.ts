import { getDataloaders } from "../graphql/loaderRegister";

require("dotenv").config();
import Koa from "koa";
import GraphQLHTTP from "koa-graphql";
import Router from "koa-router";
import cors from "@koa/cors";
import bodyParser from "koa-bodyparser";

import { getUser } from "./getUser";
import { schema } from "./schema";

import { auth } from "./auth";

const app = new Koa();
const router = new Router();

router.use(auth);

const graphqlSettingsPerReq = async (req, ctx, koaContext) => {
  const { user, team } = koaContext;
  const dataloaders = getDataloaders();

  return {
    graphiql: process.env.NODE_ENV !== "production",
    schema,
    context: {
      user,
      team,
      req,
      dataloaders,
    },
    formatError: (error) => {
      console.log(error.message);
      console.log(error.locations);
      console.log(error.stack);

      return {
        message: error.message,
        locations: error.locations,
        stack: error.stack,
      };
    },
  };
};

const graphqlServer = GraphQLHTTP(graphqlSettingsPerReq);

router.all("/graphql", graphqlServer);

app.use(bodyParser());
app.use(cors());
app.use(router.routes()).use(router.allowedMethods());

export default app;
