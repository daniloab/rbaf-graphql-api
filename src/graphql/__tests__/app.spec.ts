import request from "supertest";

import app from "../app";
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
} from "../../../test";
import { generateToken } from "../auth";
import { createUser } from "../modules/user/fixture/createUser";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it("should return 200 and return logged user", async () => {
  const user = await createUser({
    name: "Danilo",
  });

  const token = generateToken(user);

  // language=GraphQL
  const query = `
    query Q {
      me {
        name
      }      
    }
  `;

  const variables = {};

  const payload = {
    query,
    variables,
  };

  const response = await request(app.callback())
    .post("/graphql")
    .set({
      Accept: "application/json",
      "Content-Type": "application/json",
      Authorization: token,
    })
    .send(JSON.stringify(payload));

  expect(response.status).toBe(200);

  expect(response.body.data.me.name).toBe(user.name);

  expect(response.body).toMatchSnapshot();
});
