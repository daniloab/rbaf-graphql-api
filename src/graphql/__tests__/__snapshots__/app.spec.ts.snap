// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return 200 and return logged user 1`] = `
Object {
  "data": Object {
    "me": Object {
      "name": "Danilo",
    },
  },
}
`;
